File: creational/singleton/java/realLife/SingletonJavaRealWorld/src/EventService.java
import java.util.ArrayList;
import java.util.List;
public class EventService {
  private static EventService instance;

  private EventService() {}

  public static EventService getInstance() {
    if(instance == null) {
      instance = new EventService();
    }
    return instance;
  }
  
  private final List<Listener> listeners = new ArrayList<>();
  
  public void emit(String value) {
    notifyListeners(value);
  }
  
  public Unsubscribe subscribe(Listener listener) {
    listeners.add(listener);
    return new Unsubscribe(() -> listeners.remove(listener));
  }
  
  public void notifyListeners(String value) {
    for(Listener listener : listeners) {
      listener.invoke(value);
    }
  }
}

File: design-patterns/creational/singleton/swift/realLife/Sources/singletonRealLife/EventService.swift
import Foundation;

struct Unsubscribe {
    var unsubscribe: () -> Void
}

typealias Listener = (_ value: String) -> Void


class EventService {
  static var shared: EventService = {
    let instance: EventService = EventService();
    return instance
  }()

  private init () {}

  private var listeners = Dictionary<String, Listener>();
}

extension EventService: NSCopying {
  func copy(with zone: NSZone? = nil) -> Any {
    return self
  }
}

// Business Logic

extension EventService {
  public func emit(value: String) -> Void {
    self.notifyListeners(with: value);
  }

  func subscribe(_ listener: @escaping Listener) -> Unsubscribe {
    let key = UUID().uuidString;
    listeners[key] = listener;

    return Unsubscribe {
      self.listeners.removeValue(forKey: key)
    }
  }

  private func notifyListeners(with value: String) {
    for listener in self.listeners {
      listeners[listener.key]?(value)
    }
  }
}
